File: Data Structures & Algorithms/longest-increasing-subsequence/submission-2.py
class Solution:
    def lengthOfLIS(self, nums: List[int]) -> int:
        def dfs(i):
            if i == len(nums)-1:
                return 1
            if i in memo:
                return memo[i]
            
            length = 1
            for j in range(i+1, len(nums)):
                if nums[j] > nums[i]:
                    length = 1
                    length += dfs(j)
                memo[i] = max(memo.get(i,0), length)

            return memo[i]

        memo = {}
        res = 0
        for i in range(len(nums)):
            res = max(res, dfs(i))
        return res

File: Data Structures & Algorithms/combination-target-sum/submission-2.cs
public class Solution {
    private void DFS(int i, int total, int target, int[] nums, List<int> curr, List<List<int>> res){
        if (total == target){
            res.Add(new List<int>(curr));
            return;
        }

        for (int j = i; j < nums.Length; j++){
            if (total + nums[j] > target) return;

            curr.Add(nums[j]);
            DFS(j, total+nums[j], target, nums, curr, res);
            curr.RemoveAt(curr.Count-1);
        }
    }

    public List<List<int>> CombinationSum(int[] nums, int target) {
        var res = new List<List<int>>();
        var curr = new List<int>();

        Array.Sort(nums);
        DFS(0, 0, target, nums, curr, res);
        return res; 
    }
}
